Resources/doc/configuration.md
# Configuration tree

Here's an example of full configuration with all possible options including default values:

```yml
ongr_elasticsearch:
    analysis:
        analyzer:
            pathAnalyzer:
                type: custom
                tokenizer: pathTokenizer
        tokenizer:
            pathTokenizer:
                type : path_hierarchy
                buffer_size: 2024
                skip: 0
                delimiter: /
        filter:
            incremental_filter:
                type: edge_ngram
                min_gram: 1
                max_gram: 20
    managers:
        default:
            index: 
                hosts:
                    - 127.0.0.1:9200
                index_name: ongr-default
                settings:
                    refresh_interval: -1
                    number_of_replicas: 0
                    number_of_shards: 1
            logger: true #default %kernel.debug%
            mappings:
                - AcmeBarBundle #Scans all bundle documents
        custom:
            index: 
                hosts:
                    - 10.0.0.1:9200 #default 127.0.0.1:9200
                index_name: ongr-custom
                mappings:
                    AcmeBundle:
                        document_dir: Document
```

Advanced `hosts` configuration (as seen in the [Extended Host Configuration](https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/_configuration.html#_extended_host_configuration) of the Elasticsearch-PHP configuration guide):
```yml
ongr_elasticsearch:
    managers:
        default:
            index:
                hosts:
                    - 127.0.0.1:9200
                    - https://username:password@asdzxc-456123.eu-west-1.bonsaisearch.net
                    -
                        host: 127.0.0.1
                        port: 9200
                    -
                        host: asdzxc-456123.eu-west-1.bonsaisearch.net
                        port: 443
                        scheme: https
                        user: username
                        pass: password
```


DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the ONGR package.
 *
 * (c) NFQ Technologies UAB <info@nfq.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace ONGR\ElasticsearchBundle\DependencyInjection;

use Psr\Log\LogLevel;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from app/config files.
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('ongr_elasticsearch');

        $rootNode
            ->children()
            ->booleanNode('cache')
                ->info(
                    'Enables cache handler to store metadata and other data to the cache. '.
                    'By default it is enabled in prod environment and disabled in dev.'
                )
            ->end()
            ->booleanNode('profiler')
                ->info(
                    'Enables ElasticsearchBundle query profiler. Default value is kernel.debug parameter. '.
                    'If profiler is disabled the tracer service will be disabled as well.'
                )
            ->end()
            ->append($this->getAnalysisNode())
            ->append($this->getManagersNode())
            ->end();

        return $treeBuilder;
    }

    /**
     * Analysis configuration node.
     *
     * @return \Symfony\Component\Config\Definition\Builder\NodeDefinition
     */
    private function getAnalysisNode()
    {
        $builder = new TreeBuilder();
        $node = $builder->root('analysis');

        $node
            ->info('Defines analyzers, tokenizers and filters')
            ->addDefaultsIfNotSet()
            ->children()
                ->arrayNode('tokenizer')
                    ->defaultValue([])
                    ->prototype('variable')->end()
                ->end()
                ->arrayNode('filter')
                    ->defaultValue([])
                    ->prototype('variable')->end()
                ->end()
                ->arrayNode('analyzer')
                    ->defaultValue([])
                    ->prototype('variable')->end()
                ->end()
                ->arrayNode('char_filter')
                    ->defaultValue([])
                    ->prototype('variable')->end()
                ->end()
            ->end();

        return $node;
    }

    /**
     * Managers configuration node.
     *
     * @return \Symfony\Component\Config\Definition\Builder\NodeDefinition
     */
    private function getManagersNode()
    {
        $builder = new TreeBuilder();
        $node = $builder->root('managers');

        $node
            ->isRequired()
            ->requiresAtLeastOneElement()
            ->useAttributeAsKey('name')
            ->info('Maps managers to connections and bundles')
            ->prototype('array')
                ->children()
                    ->arrayNode('index')
                        ->children()
                            ->scalarNode('index_name')
                                ->isRequired()
                                ->info('Sets index name for connection.')
                            ->end()
                            ->arrayNode('hosts')
                                ->info('Defines hosts to connect to.')
                                ->defaultValue(['127.0.0.1:9200'])
                                ->prototype('array')
                                    ->beforeNormalization()
                                    ->ifString()
                                        ->then(function($v) {
                                            $parts = parse_url($v);
                                            if (!$parts['port'] && !!$parts['scheme']) {
                                                switch ($parts['scheme']) {
                                                    case 'https':
                                                        $parts['port'] = 443;
                                                        break;
                                                    case 'http':
                                                        $parts['port'] = 80;
                                                        break;
                                                }
                                            }
                                            return $parts;
                                        })
                                    ->end()
                                    ->children()
                                        ->scalarNode('host')->isRequired()->end()
                                        ->scalarNode('port')->end()
                                        ->scalarNode('scheme')->end()
                                        ->scalarNode('user')->end()
                                        ->scalarNode('pass')->end()
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('settings')
                                ->defaultValue(
                                    [
                                        'number_of_replicas' => 0,
                                        'number_of_shards' => 1,
                                        'refresh_interval' => -1,
                                    ]
                                )
                                ->info('Sets index settings for connection.')
                                ->prototype('variable')->end()
                            ->end()
                        ->end()
                    ->end()
                    ->integerNode('bulk_size')
                        ->min(0)
                        ->defaultValue(100)
                        ->info(
                            'Maximum documents size in the bulk container. ' .
                            'When the limit is reached it will auto-commit.'
                        )
                    ->end()
                    ->enumNode('commit_mode')
                        ->values(['refresh', 'flush', 'none'])
                        ->defaultValue('refresh')
                        ->info(
                            'The default type of commit for bulk queries.'
                        )
                    ->end()
                    ->arrayNode('logger')
                        ->info('Enables elasticsearch queries logging')
                        ->addDefaultsIfNotSet()
                        ->beforeNormalization()
                            ->ifTrue(
                                function ($v) {
                                    return is_bool($v);
                                }
                            )
                            ->then(
                                function ($v) {
                                    return ['enabled' => $v];
                                }
                            )
                        ->end()
                        ->children()
                            ->booleanNode('enabled')
                                ->info('enables logging')
                                ->defaultFalse()
                            ->end()
                            ->scalarNode('level')
                                ->info('Sets PSR logging level.')
                                ->defaultValue(LogLevel::WARNING)
                                ->validate()
                                ->ifNotInArray((new \ReflectionClass('Psr\Log\LogLevel'))->getConstants())
                                    ->thenInvalid('Invalid PSR log level.')
                                ->end()
                            ->end()
                            ->scalarNode('log_file_name')
                                ->info('Log filename. By default it is a manager name.')
                                ->defaultValue(null)
                            ->end()
                        ->end()
                    ->end()
                    ->arrayNode('mappings')
                        ->info('Maps manager to the bundles. f.e. AppBundle')
                        ->prototype('variable')->end()
                    ->end()
                    ->booleanNode('force_commit')
                        ->info('Forces commit to the elasticsearch on kernel terminate event.')
                        ->defaultTrue()
                    ->end()
                ->end()
            ->end();

        return $node;
    }
}
